File: frontend/src/lib/api/plans.ts
/**
 * Make It Work plan endpoints — narrative, email-first save, magic-link consume.
 */

import { api } from '@/lib/api-client'
import type { ScenarioPayloadV1 } from '@/lib/dealStructures/scenarioPayload'
import type { WizardAnswers } from '@/components/iq-verdict/make-it-work/wizardMapping'

export interface PlanNarrativeRequest {
  address: string
  family: string
  family_label: string
  headline: string
  bullets: string[]
  levers: Array<{ label: string; before_label: string; after_label: string }>
  monthly_savings: number
  cash_required: number
  list_price: number | null
  target_buy_price: number | null
  wizard_answers: WizardAnswers
}

export interface PlanNarrative {
  summary: string
  pitch: string
  source: 'ai' | 'template'
}

export function requestPlanNarrative(body: PlanNarrativeRequest): Promise<PlanNarrative> {
  return api.post<PlanNarrative>('/api/v1/plans/narrative', body, { softAuth: true, timeoutMs: 12_000 })
}

export interface BreakevenWayInput {
  family: 'price' | 'income' | 'financing' | 'capital_stack'
  name: string
  change_pct: number | null
  change_amount: number | null
  result_amount: number
  result_label: string
  closes_gap_alone: boolean
  terms_note: string | null
  rating: 'high' | 'medium' | 'low' | 'your_call' | null
  reasons: string[]
  cash_required: number | null
}

export interface BreakevenNarrativeRequest {
  address: string
  list_price: number | null
  target_buy_price: number | null
  income_value: number | null
  gap_amount: number | null
  gap_pct: number | null
  monthly_shortfall: number | null
  baseline_cash_required: number | null
  ways: BreakevenWayInput[]
  blend_recommendation: string | null
}

/**
 * The section's "Your move". Deliberately not per-lever text: every row
 * already shows its own change, result, and likelihood, so prose that repeats
 * them adds nothing. This is sequencing and a walk-away instead.
 */
export interface BreakevenNarrative {
  /** The opening play and the order of asks. */
  move: string
  /** The line past which this stops being a deal. */
  walk_away: string
  source: 'ai' | 'template'
}

export function requestBreakevenNarrative(body: BreakevenNarrativeRequest): Promise<BreakevenNarrative> {
  return api.post<BreakevenNarrative>('/api/v1/plans/breakeven-narrative', body, {
    softAuth: true,
    timeoutMs: 12_000,
  })
}

export interface PlanClaimRequest {
  email: string
  address: string
  address_parts: {
    street: string
    city?: string
    state?: string
    zip?: string
  }
  zpid?: string | null
  latitude?: number | null
  longitude?: number | null
  property_snapshot: Record<string, unknown>
  scenario: ScenarioPayloadV1 | null
  wizard_answers: WizardAnswers
  narrative: PlanNarrative | null
}

export interface PlanClaimResponse {
  status: 'accepted'
  message: string
}

export function claimPlan(body: PlanClaimRequest): Promise<PlanClaimResponse> {
  return api.post<PlanClaimResponse>('/api/v1/plans/claim', body, { softAuth: true })
}

export interface MagicLinkConsumeResponse {
  redirect: string
  access_token?: string | null
  refresh_token?: string | null
}

export function consumeMagicLink(
  token: string,
  next?: string | null,
): Promise<MagicLinkConsumeResponse> {
  const query = next ? `?${new URLSearchParams({ next }).toString()}` : ''
  return api.post<MagicLinkConsumeResponse>(
    `/api/v1/auth/magic-link/consume${query}`,
    { token },
    { skipAuth: true },
  )
}

File: backend/app/schemas/plans.py
"""Schemas for the Make It Work plan endpoints (narrative, claim, magic link)."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, EmailStr, Field, field_validator


class WizardAnswers(BaseModel):
    cash: str | None = None
    priority: str | None = None
    terms: str | None = None
    owner_occupy: bool | None = Field(None, alias="ownerOccupy")

    model_config = {"populate_by_name": True}


class PlanLever(BaseModel):
    label: str = Field(..., max_length=80)
    before_label: str = Field("", max_length=80)
    after_label: str = Field("", max_length=120)


class PlanNarrativeRequest(BaseModel):
    address: str = Field(..., max_length=300)
    family: str = Field(..., max_length=30)
    family_label: str = Field(..., max_length=80)
    headline: str = Field(..., max_length=200)
    bullets: list[str] = Field(default_factory=list, max_length=6)
    levers: list[PlanLever] = Field(default_factory=list, max_length=6)
    monthly_savings: float = 0.0
    cash_required: float = 0.0
    list_price: float | None = None
    target_buy_price: float | None = None
    wizard_answers: WizardAnswers = Field(default_factory=WizardAnswers)

    @field_validator("bullets")
    @classmethod
    def _trim_bullets(cls, v: list[str]) -> list[str]:
        return [b[:200] for b in v]


class PlanNarrativeResponse(BaseModel):
    summary: str
    pitch: str
    source: Literal["ai", "template"]


class BreakevenWayInput(BaseModel):
    """One row of the Breakeven Analysis section, as the frontend already shows it."""

    family: Literal["price", "income", "financing", "capital_stack"]
    name: str = Field(..., max_length=40)
    change_pct: float | None = None
    change_amount: float | None = None
    result_amount: float
    result_label: str = Field(..., max_length=40)
    closes_gap_alone: bool = True
    terms_note: str | None = Field(None, max_length=160)
    rating: Literal["high", "medium", "low", "your_call"] | None = None
    reasons: list[str] = Field(default_factory=list, max_length=8)
    cash_required: float | None = Field(None, description="Cash to close under this lever")

    @field_validator("reasons")
    @classmethod
    def _trim_reasons(cls, v: list[str]) -> list[str]:
        return [r[:200] for r in v]


class BreakevenNarrativeRequest(BaseModel):
    address: str = Field(..., max_length=300)
    list_price: float | None = None
    target_buy_price: float | None = None
    income_value: float | None = None
    gap_amount: float | None = None
    gap_pct: float | None = None
    monthly_shortfall: float | None = None
    baseline_cash_required: float | None = None
    ways: list[BreakevenWayInput] = Field(default_factory=list, max_length=4)
    blend_recommendation: str | None = Field(None, max_length=600)


class BreakevenNarrativeResponse(BaseModel):
    """The one thing the numbers can't supply: what to actually do.

    Deliberately not a per-lever paragraph. Each row already shows its own
    change, result, and likelihood; restating those in prose is what made the
    section read like notes. This is sequencing and a walk-away instead.
    """

    move: str = Field(..., description="The opening play and the order of asks")
    walk_away: str = Field(..., description="The line past which this stops being a deal")
    source: Literal["ai", "template"]


class PlanScenario(BaseModel):
    """Mirror of the frontend ``ScenarioPayloadV1``."""

    v: int = 1
    structure_id: str = Field(..., alias="structureId", max_length=80)
    family: str = Field(..., max_length=30)
    label: str = Field("", max_length=120)
    levers: dict[str, Any] = Field(default_factory=dict)

    model_config = {"populate_by_name": True}


class PlanAddressParts(BaseModel):
    street: str = Field(..., max_length=255)
    city: str | None = Field(None, max_length=100)
    state: str | None = Field(None, max_length=10)
    zip: str | None = Field(None, max_length=20)


class PlanClaimRequest(BaseModel):
    email: EmailStr
    address: str = Field(..., max_length=500)
    address_parts: PlanAddressParts
    zpid: str | None = Field(None, max_length=50)
    latitude: float | None = Field(None, ge=-90, le=90)
    longitude: float | None = Field(None, ge=-180, le=180)
    property_snapshot: dict[str, Any] = Field(default_factory=dict)
    scenario: PlanScenario | None = None
    wizard_answers: WizardAnswers = Field(default_factory=WizardAnswers)
    narrative: PlanNarrativeResponse | None = None


class PlanClaimResponse(BaseModel):
    status: Literal["accepted"] = "accepted"
    message: str = "If that address is valid, your plan is on its way."


class MagicLinkConsumeRequest(BaseModel):
    token: str = Field(..., min_length=16, max_length=256)


class MagicLinkConsumeResponse(BaseModel):
    redirect: str
    access_token: str | None = None
    refresh_token: str | None = None
